include/verilated_saif_c.h

```diff
@@ -0,0 +1,430 @@
+
+    // METHODS
+    VL_ATTR_ALWINLINE void addChildScopeIndex(int32_t index) {
+        m_childScopesIndices.emplace_back(index);
+    }
+    VL_ATTR_ALWINLINE void addActivityVar(uint32_t code, std::string name) {
+        m_childActivities.emplace_back(code, std::move(name));
+    }
+    VL_ATTR_ALWINLINE bool hasParent() const { return m_parentScopeIndex >= 0; }
+
+    // ACCESSORS
+    VL_ATTR_ALWINLINE const std::string& getName() const { return m_scopeName; }
+    VL_ATTR_ALWINLINE const std::vector<int32_t>& getChildScopesIndices() const {
+        return m_childScopesIndices;
+    }
+    VL_ATTR_ALWINLINE const std::vector<std::pair<uint32_t, std::string>>&
+    getChildActivities() const {
+        return m_childActivities;
+    }
+    VL_ATTR_ALWINLINE int32_t getParentScopeIndex() const { return m_parentScopeIndex; }
+
+private:
+    // CONSTRUCTORS
+    VL_UNCOPYABLE(ActivityScope);
+
+    // MEMBERS
+    std::string m_scopeName{};
+    std::vector<int32_t> m_childScopesIndices{};
+    std::vector<std::pair<uint32_t, std::string>> m_childActivities{};
+    int32_t m_parentScopeIndex{-1};
+};
+
+//=============================================================================
+// VerilatedSaif
+// Base class to create a Verilator SAIF dump
+// This is an internally used class - see VerilatedSaifC for what to call from applications
+
+class VerilatedSaif VL_NOT_FINAL : public VerilatedTrace<VerilatedSaif, VerilatedSaifBuffer> {
+public:
+    using Super = VerilatedTrace<VerilatedSaif, VerilatedSaifBuffer>;
+
+private:
+    friend VerilatedSaifBuffer;  // Give the buffer access to the private bits
+
+    //=========================================================================
+    // SAIF-specific internals
+
+    VerilatedSaifFile* m_filep;  // File we're writing to
+    bool m_fileNewed;  // m_filep needs destruction
+    bool m_isOpen = false;  // True indicates open file
+    std::string m_filename;  // Filename we're writing to (if open)
+    uint64_t m_rolloverSize = 0;  // File size to rollover at
+
+    void initializeSaifFileContents();
+    void finalizeSaifFileContents();
+    void recursivelyPrintScopes(uint32_t scopeIndex);
+    void openInstanceScope(const std::string& instanceName);
+    void closeInstanceScope();
+    void printScopeActivities(const ActivityScope& scope);
+    void openNetScope();
+    void closeNetScope();
+    bool printActivityStats(uint32_t activityCode, const char* activityName, bool anyNetValid);
+
+    void incrementIndent();
+    void decrementIndent();
+    void printIndent();
+
+    int m_indent = 0;
+
+    void printStr(const char* str);
+    void printStr(const std::string& str);
+
+    void clearCurrentlyCollectedData();
+
+    int32_t m_currentScope{-1};
+    std::vector<ActivityScope> m_scopes{};
+    std::vector<int32_t> m_topScopes{};
+
+    std::unordered_map<uint32_t, ActivityVar> m_activity;
+    std::vector<std::vector<ActivityBit>> m_activityArena;
```

Comment:
I suggest abstracting these into a separate ActivityAccumulator, or whatever class, so you can have multiple independent accumulators, then you can use the same style of parallelism as in #5806, where each thread collects into a separate accumulators, then you can combine all accumulators when at the end of simulation when you actually write the file.

At the same time, move all the `Activity*` classes into verilated_saif_c.c, and in this header, just hold forward declared pointers to them. This header is included in every verilator output file, so it would be good to keep it smaller.